bson/src/main/org/bson/Document.java

```diff
@@ -158,6 +162,79 @@ public Document append(final String key, final Object value) {
+     *
+     * @param keys  the list of keys
+     * @param clazz the non-null class to cast the value to
+     * @param <T>   the type of the class
+     * @return the value of the given embedded key, or null if the instance does not contain this embedded key.
+     * @throws ClassCastException if the value of the given embedded key is not of type T
+     * @since 3.10
+     */
+    public <T> T getEmbedded(final List<?> keys, final Class<T> clazz) {
+        notNull("keys", keys);
+        isTrue("keys", !keys.isEmpty());
+        notNull("clazz", clazz);
+        return getEmbeddedValue(keys, clazz, null);
+    }
+
+    /**
+     * Gets the value in an embedded document, casting it to the given {@code Class<T>} or returning the default value if null.
+     * The list of keys represents a path to the embedded value, drilling down into an embedded document for each key.
+     * This is useful to avoid having casts in client code, though the effect is the same.
+     *
+     * The generic type of the keys list is {@code ?} to be consistent with the corresponding {@code get} methods, but in practice
+     * the actual type of the argument should be {@code List<String>}. So to get the embedded value of a key list that is of type String,
+     * you would write {@code String name = doc.getEmbedded(List.of("employee", "manager", "name"), "John Smith")} instead of
+     * {@code String name = doc.get("employee", Document.class).get("manager", Document.class).get("name", "John Smith") }.
+     *
+     * @param keys  the list of keys
+     * @param defaultValue what to return if the value is null
+     * @param <T>   the type of the class
+     * @return the value of the given key, or null if the instance does not contain this key.
+     * @throws ClassCastException if the value of the given key is not of type T
+     * @since 3.10
+     */
+    public <T> T getEmbedded(final List<?> keys, final T defaultValue) {
+        notNull("keys", keys);
+        isTrue("keys", !keys.isEmpty());
+        notNull("defaultValue", defaultValue);
+        return getEmbeddedValue(keys, null, defaultValue);
+    }
+
+
+    // Gets the embedded value of the given list of keys, casting it to {@code Class<T>} or returning the default value if null.
+    // Throws ClassCastException if any of the intermediate embedded values is not a Document.
+    @SuppressWarnings("unchecked")
+    private <T> T getEmbeddedValue(final List<?> keys, final Class<T> clazz, final T defaultValue) {
+        Object value = this;
+        Iterator<?> keyIterator = keys.iterator();
+        while (keyIterator.hasNext()) {
+            Object key = keyIterator.next();
+            value = ((Document) value).get(key);
+            if (!(value instanceof Document)) {
+                if (value == null) {
+                    if (keyIterator.hasNext()) {
+                        throw new ClassCastException(format("The key %s is not a Document", key));
```

Comment:
We agreed in Slack that:

    Document.parse("{}").getEmbedded(List.of("a", "b"), Integer.class)

should return null, not throw ClassCastException.  But that's not what's happening here.

Reply:
Yes we did. An oversight on my part. I have added this example as a test.

Reply:
I wasn't clear about my intention with that example.  It's not about whether the top-level document is empty, it's about whether the requested key exists in the document at the current level.  So all of these should return null (or the default value):

        Document.parse('{}').getEmbedded(['a'], Integer)
        Document.parse('{b: 1}').getEmbedded(['a'], Integer)
        Document.parse('{b: 1}').getEmbedded(['a', 'b'], Integer)
        Document.parse('{a: {c: 1}}').getEmbedded(['a', 'b'], Integer)
        Document.parse('{a: {c: 1}}').getEmbedded(['a', 'b', 'c'], Integer)

Reply:
It turns out that I originally had this functionality and changed it. I have added your examples to the tests for null. All tests pass. `./gradlew check` successfully runs with the latest change.